src/wallet.rs

```diff
@@ -194,6 +205,175 @@ impl Wallet {
     }
 }
 
+#[derive(Clone, Copy, Debug)]
+pub struct ShardConfig {
+    /// Number of shards to break the key into
+    pub key_share_count: u8,
+
+    /// Number of shards required to recover the key
+    pub recovery_threshold: u8,
+}
+
+pub struct Builder {
+    /// Output file to store the key in
+    output: PathBuf,
+
+    /// Password to access wallet
+    password: String,
+
+    /// Overwrite an existing file
+    force: bool,
+
+    /// Use a BIP39 or mobile app seed phrase to generate the wallet keys
+    seed_type: Option<SeedType>,
+
+    /// The seed words used to create this wallet
+    seed_words: Option<Vec<String>>,
+
+    /// The KeyTag (network and key type) to use for this wallet
+    key_tag: KeyTag,
+
+    /// Optional shard config info to use in order to create a sharded wallet
+    /// otherwise, creates a basic non-sharded wallet
+    shard_config: Option<ShardConfig>,
+}
+
+impl Builder {
+    pub fn new() -> Builder {
+        Builder {
+            output: PathBuf::from("wallet.key"),
+            password: Default::default(),
+            force: false,
+            seed_type: None,
+            seed_words: None,
+            key_tag: Default::default(),
+            shard_config: None,
+        }
+    }
+
+    /// Sets the output file for the wallet.
+    /// Defaults to 'main.key'
+    pub fn output(mut self, path: &Path) -> Builder {
```

Comment:
I'm wondering why you're not taking `&mut self`? This "by value" approach would be copying self often?